### DISCUSSION TITLE
FrankenPHP docker container with Laravel crashes with Code 139 (SIGSEGV Segmentation Fault)

### DISCUSSION THREAD

**Discussion by Kovah** [php/frankenphp#321] [Q&A]
I recently tried out FrankenPHP with a pretty basic Laravel application. While the local development environment is working fine, the production deployment to a Proxmox VM running Docker crashes with Code 139 (SIGSEGV Segmentation Fault) seemingly random, but after only a couple of requests. Unfortunately, there are no errors logged, neither in the container nor on the host.

**Dockerfile**
```Dockerfile
FROM dunglas/frankenphp:latest-php8.2 as app

RUN install-php-extensions opcache zip bcmath intl

COPY docker/php/php-prod.ini /usr/local/etc/php/conf.d/app.ini
COPY ./docker/caddy/Caddyfile /etc/caddy/Caddyfile

COPY app /app/app
COPY bootstrap /app/bootstrap
COPY config /app/config
COPY database /app/database
COPY lang /app/lang
COPY public /app/public
COPY resources /app/resources
COPY routes /app/routes
COPY --chown=www-data:www-data storage /app/storage

COPY artisan composer.json composer.lock README.md /app/

COPY --from=composer:latest /usr/bin/composer /usr/bin/composer
RUN /usr/bin/composer install --no-interaction --no-dev --optimize-autoloader \
    && /usr/bin/composer clear-cache

# Vite assets only from a previous stage
COPY --from=frontend-builder /app/public/build /app/public/build
```

**Caddyfile**
```
{
    debug
    frankenphp
    order php_server before file_server
}
:8080 {
    root * /app/public
    php_server
}
```
From my point of view, there's nothing really special with the setup. Caddy listens on port 8080 without a host because HTTPS is terminated on a load balancer in front of the container.

How could I debug this further and find out the root cause for this crash? As mentioned before, there are no logs from the container.

---

**Update:**

Found #262 where using gdb was suggested. I tinkered a bit and got it working with `gdb -ex run --args frankenphp "run" "--config" "/etc/caddy/Caddyfile" "--adapter" "caddyfile"`. FrankenPHP is starting up and the first request came in, but there's no stack trace, it only crashes with
```
2023/11/28 10:42:31.134	DEBUG	http.handlers.rewrite	rewrote request	{"request": {"remote_ip": "192.168.65.1", "remote_port": "65502", .....}, "method": "GET", "uri": "/index.php"}

Thread 12 "thpool-0" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0xffffa276dde0 (LWP 947)]
```

**Comment by dunglas** [comment]
Thanks for the report! Can you run the `bt` (for backtrace) command in the GDB shell after the crash and copy the output? This should give a stack trace.

Another option is to get a core dump: https://stackoverflow.com/a/47694315/1352334

**Reply by Kovah** [reply]
Unfortunately, I only get this when running bt:
```
#0  0x0000ffff6dc00e0c in ?? ()
#1  0x0000ffff88413020 in ?? ()
Backtrace stopped: previous frame inner to this frame (corrupt stack?)
```

I was not able to get any core dumps and I am afraid to tinker with my macOS system to get this working in Docker. It seems to be not a common use case to access core dumps of a Docker container on macOS. All tutorials I tried did either not work or looked scary to try.

**Reply by withinboredom** [reply]
@Kovah is it using Fibers?

**Reply by dunglas** [reply]
The problem with Fibers is usually not a segmentation fault.

@Kovah is it possible to access to repo with a reproducer?

**Reply by Kovah** [reply]
The repo is a company one, so no. But I'll see if I can reproduce this with a plain Laravel project. Maybe I also find the package/library which causes this.

**Reply by IgordeOliveira** [reply]
While attempting to create my development environment, I encountered the same issue. It all started when I followed Laravel's recommendation to reload works after file changes using the node chokidar library (https://laravel.com/docs/10.x/octane#watching-for-file-changes). Whenever I change any file, the Docker container crashes and returns an error code 139.

Here is my dev docker file:
```
FROM dunglas/frankenphp:latest-php8.3
RUN install-php-extensions gd intl exif pcntl pdo_pgsql gmp opcache zip
RUN curl -fsSL https://deb.nodesource.com/setup_21.x | bash - \
    && apt-get install -y nodejs \
    && npm install -g chokidar

```
Also, my docker-compose 
```
    php:
      build:
        context: .
        dockerfile: dev.Dockerfile
      entrypoint: php artisan octane:start --watch --host=0.0.0.0
      ports:
        - "80:8000"
      volumes:
        - .:/app
      tty: true
```

**Reply by GhostvOne** [reply]
I'm having excatly the same issue
